node/pkg/tss/util.go

```diff
@@ -0,0 +1,212 @@
+		l.Error(err.Error())
+		return
+	}
+
+	switch {
+	case informativeErr.round != "", informativeErr.trackingId != nil:
+		l.Error(
+			informativeErr.Error(),
+			zap.String("round", string(informativeErr.round)),
+			zap.String("trackingId", fmt.Sprintf("%x", informativeErr.trackingId)),
+		)
+	case informativeErr.trackingId != nil:
+		l.Error(
+			informativeErr.Error(),
+			zap.String("trackingId", fmt.Sprintf("%x", informativeErr.trackingId)),
+		)
+	case informativeErr.round != "":
+		l.Error(
+			informativeErr.Error(),
+			zap.String("round", string(informativeErr.round)),
+		)
+	default:
+		// no additional fields
+		l.Error(informativeErr.Error())
+	}
+}
+
+func equalPartyIds(a, b *tss.PartyID) bool {
+	return a.Id == b.Id && string(a.Key) == string(b.Key)
+}
+
+func protoToPartyId(pid *tsscommv1.PartyId) *tss.PartyID {
+	return &tss.PartyID{
+		MessageWrapper_PartyID: &tss.MessageWrapper_PartyID{
+			Id:      pid.Id,
+			Moniker: pid.Moniker,
+			Key:     pid.Key,
+		},
+		Index: int(pid.Index),
+	}
+}
+
+func partyIdToProto(pid *tss.PartyID) *tsscommv1.PartyId {
+	return &tsscommv1.PartyId{
+		Id:      pid.Id,
+		Moniker: pid.Moniker,
+		Key:     pid.Key,
+		Index:   uint32(pid.Index),
+	}
+}
+
+var (
+	ErrEchoIsNil             = fmt.Errorf("echo is nil")
+	ErrNoEchoSignature       = fmt.Errorf("echo doesn't contain a signature")
+	ErrNoAuthenticationField = fmt.Errorf("SignedMessage doesn't contain an authentication field")
+	ErrNilPartyId            = fmt.Errorf("party id is nil")
+	ErrEmptyIDInPID          = fmt.Errorf("partyId identifier is empty")
+	ErrEmptyKeyInPID         = fmt.Errorf("partyId doesn't contain a key")
+	ErrSignedMessageIsNil    = fmt.Errorf("SignedMessage is nil")
+	ErrNilPayload            = fmt.Errorf("SignedMessage doesn't contain a payload")
+)
+
+func vaidateEchoCorrectForm(e *tsscommv1.Echo) error {
+	if e == nil {
+		return ErrEchoIsNil
+	}
+
+	if err := validatePartIdProtoCorrectForm(e.Echoer); err != nil {
+		return err
+	}
+
+	if err := validateSignedMessageCorrectForm(e.Message); err != nil {
+		return fmt.Errorf("echo message error:%w", err)
+	}
+
+	if len(e.Message.Signature) == 0 {
+		return ErrNoAuthenticationField
+	}
+
+	return nil
+}
+
+func validatePartIdProtoCorrectForm(p *tsscommv1.PartyId) error {
+	if p == nil {
+		return ErrNilPartyId
+	}
+
+	if p.Id == "" {
+		return ErrEmptyIDInPID
+	}
+
+	if len(p.Key) == 0 {
+		return ErrEmptyKeyInPID
+	}
+
+	return nil
+
+}
+
+func validateSignedMessageCorrectForm(m *tsscommv1.SignedMessage) error {
+	if m == nil {
+		return ErrSignedMessageIsNil
+	}
+
+	if err := validatePartIdProtoCorrectForm(m.Sender); err != nil {
+		return fmt.Errorf("signedMessage sender pID error:%w", err)
+	}
+
+	if m.Payload == nil {
+		return ErrNilPayload
+	}
+
+	for _, v := range m.Recipients {
+		if err := validatePartIdProtoCorrectForm(v); err != nil {
+			return err
+		}
+	}
+
+	// not checking the signature, since it is allowed to be nil in unicast.
+	return nil
```

Comment:
can you check the sig if not unicast?

Reply:
I can, though i do it outside this function. 
This function is mainly to ensure the form of the message is correct, and doesn't misses any fields. 
signature verification happens afterwards.

Reply:
I changed it once i've changed the types. I think it's better now.